### DISCUSSION TITLE
Serialization of std::map<std::pair<std::string, std::string>, std::vector<std::string>> is inconsistent?

### DISCUSSION THREAD

**Discussion by EddieDover** [nlohmann/json#4013] [Q&A] [answered]
Given the following code:
```
	std::map<std::pair<std::string, std::string>, std::vector<std::string> > holder;

	holder[std::make_pair("key_a1", "key_a2")] = std::vector<std::string>({"value_a1", "value_a2"});
	holder[std::make_pair("key_b1", "key_b2")] = std::vector<std::string>({"value_b1", "value_b2", "value_b3"});
	holder[std::make_pair("key_c1", "key_c2")] = std::vector<std::string>({"value_c1", "value_c2", "value_c3", "value_c4"});

	json j = holder;
	std::ofstream o("report.json");
	o << std::setw(4) << j << std::endl;
	o.close();
```
Why does the output change if the value of the set only has two items?
```
[
    {
        "key_a1": "key_a2",
        "value_a1": "value_a2"
    },
    [
        [
            "key_b1",
            "key_b2"
        ],
        [
            "value_b1",
            "value_b2",
            "value_b3"
        ]
    ],
    [
        [
            "key_c1",
            "key_c2"
        ],
        [
            "value_c1",
            "value_c2",
            "value_c3",
            "value_c4"
        ]
    ]
]

```

I even inserted the two items for A one at a time with push_back, but still saw the same result.

Is this a bug, or am I just doing something improperly?

**Comment by gregmarr** [comment]
Using anything other than string as the key of the map results in some interesting interpretations.  See the constructor documentation about interpreting collections as name value pairs vs plain arrays of values. https://json.nlohmann.me/api/basic_json/basic_json/

**Comment by EddieDover** [comment, marked as answer]
I think the actual issue here stems from Item 5b on https://json.nlohmann.me/api/basic_json/basic_json/

> If the list consists of pairs whose first element is a string, a JSON object value is created where the first elements of the pairs are treated as keys and the second elements are as values.

Since the keys are pairs of strings and the value is a string vector, if it has exactly 2 items, then it fulfills the above requirement and is treated differently. 

I intend to switch to another JSON library but, in the meanwhile, I simply added code during my JSON Read/Write functions that adds/removes a 3rd temporary item to any of the value vectors that only have two items.
